### PR TITLE
[JBIDE-26165] - Implement a Launcher project wizard

### PR DESCRIPTION
**Pull request by jeffmaury**
Signed-off-by: Jeff MAURY <jmaury@redhat.com>

### CODE REVIEW

File: common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+		super("Project details", "Select the mission and then choose the runtime", "main", wizard, null);
+		this.model = model;
+	}
+
+	private void loadCatalog() {
+		try {
+			WizardUtils.runInWizard(Job.create("Loading launcher catalog", monitor -> model.setCatalog(CatalogManager.getDefault().getCatalog(monitor))), getContainer());
+		} catch (InvocationTargetException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java]: * ignore or log? 
* unify in catch block with InterruptedException?

File: common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+		//use default location
+		Button buttonUseDefaultLocation = new Button(parent, SWT.CHECK);
+		buttonUseDefaultLocation.setText("Use default location");
+		GridDataFactory.fillDefaults().align(SWT.LEFT, SWT.CENTER).span(2, 1).applyTo(buttonUseDefaultLocation);
+		IObservableValue useDefaultLocationButtonObservable = WidgetProperties.selection().observe(buttonUseDefaultLocation);
+		Binding useDefaultLocationBinding = ValueBindingBuilder.bind(useDefaultLocationButtonObservable)
+				.to(BeanProperties.value(NewLauncherProjectModel.USE_DEFAULT_LOCATION_PROPERTY).observe(model)).in(dbc);
+
+		// location
+		//Widget txtLocation = createTextWidget(parent, dbc, "Location", NewLauncherProjectModel.LOCATION_PROPERTY, "Please specify a location for your project");
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java]: remove commented line?

File: common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+		// project name
+		createTextWidget(parent, dbc, "Project name", NewLauncherProjectModel.PROJECT_NAME_PROPERTY, new EclipseProjectValidator("Please specify an Eclipse project", "Project already exists"));
+
+		//use default location
+		Button buttonUseDefaultLocation = new Button(parent, SWT.CHECK);
+		buttonUseDefaultLocation.setText("Use default location");
+		GridDataFactory.fillDefaults().align(SWT.LEFT, SWT.CENTER).span(2, 1).applyTo(buttonUseDefaultLocation);
+		IObservableValue useDefaultLocationButtonObservable = WidgetProperties.selection().observe(buttonUseDefaultLocation);
+		Binding useDefaultLocationBinding = ValueBindingBuilder.bind(useDefaultLocationButtonObservable)
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java]: useless assignment to local variable

File: common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+		}
+
+	}
+	@Override
+	protected void doCreateControls(Composite parent, DataBindingContext dbc) {
+		GridDataFactory.fillDefaults().grab(true, false).align(SWT.FILL, SWT.TOP).applyTo(parent);
+		GridLayoutFactory.fillDefaults().margins(6, 6).numColumns(2).applyTo(parent);
+
+		// project name
+		createTextWidget(parent, dbc, "Project name", NewLauncherProjectModel.PROJECT_NAME_PROPERTY, new EclipseProjectValidator("Please specify an Eclipse project", "Project already exists"));
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java]: extensively long line, 191 characters wide.

File: common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+		Text txtLocation = new Text(parent, SWT.BORDER);
+		GridDataFactory.fillDefaults().align(SWT.FILL, SWT.CENTER).grab(true, false).applyTo(txtLocation);
+		IObservableValue locationTextObservable = WidgetProperties.text(SWT.Modify).observe(txtLocation);
+		ValueBindingBuilder.bind(locationTextObservable)
+				.validatingAfterGet(new MandatoryStringValidator("Please specify a location for you project"))
+				.converting(IConverter.create(String.class, IPath.class, NewLauncherProjectWizardPage::string2IPath))
+				.to(BeanProperties.value(NewLauncherProjectModel.LOCATION_PROPERTY).observe(model)).in(dbc);
+		IObservableValue locationTextEnabledObservable = WidgetProperties.enabled().observe(txtLocation);
+		Binding locationEnabledBinding = ValueBindingBuilder.bind(locationTextEnabledObservable)
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardPage.java]: useless assignment to local variable

File: common/plugins/org.jboss.tools.common.launcher.core/src/main/java/org/jboss/tools/common/launcher/core/model/CatalogManager.java [hunk 6]
```diff
@@ -12,17 +12,28 @@
...
 import java.io.IOException;
+import java.io.OutputStream;
 import java.util.HashMap;
 import java.util.Map;
 
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.Status;
+import org.jboss.tools.common.launcher.core.LauncherCoreConstants;
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.core/src/main/java/org/jboss/tools/common/launcher/core/model/CatalogManager.java]: import never used

File: common/plugins/org.jboss.tools.common.launcher.core/src/main/java/org/jboss/tools/common/launcher/core/model/CatalogManager.java [hunk 7]
```diff
@@ -70,4 +81,25 @@ private Catalog loadCatalog(String endpointURL, IProgressMonitor monitor) throws
...
+		}
+		endpointURL += LAUNCHER_ZIP_PREFIX;
+		endpointURL += '?';
+		endpointURL += LAUNCHER_ZIP_MISSION_PARAMETER_NAME + '=' + mission + '&';
+		endpointURL += LAUNCHER_ZIP_RUNTIME_PARAMETER_NAME + '=' + runtime + '&';
+		endpointURL += LAUNCHER_ZIP_RUNTIME_VERSION_PARAMETER_NAME + '=' + runtimeVersion + '&';
+		endpointURL += LAUNCHER_ZIP_PROJECT_NAME_PARAMETER_NAME + '=' + projectName + '&';
+		endpointURL += LAUNCHER_ZIP_GROUP_ID_PARAMETER_NAME + '=' + groupId + '&';
+		endpointURL += LAUNCHER_ZIP_ARTIFACT_ID_PARAMETER_NAME + '=' + artifactId + '&';
+		endpointURL += LAUNCHER_ZIP_VERSION_PARAMETER_NAME + '=' + version;
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.core/src/main/java/org/jboss/tools/common/launcher/core/model/CatalogManager.java]: having a test for CatalogManager would be great

File: common/plugins/org.jboss.tools.common.launcher.core/src/main/java/org/jboss/tools/common/launcher/core/model/Mission.java [hunk 8]
```diff
@@ -47,10 +47,37 @@ public void setName(String name) {
...
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((id == null) ? 0 : id.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.core/src/main/java/org/jboss/tools/common/launcher/core/model/Mission.java]: simple test for this method would be great

File: common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectModel.java [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+import java.util.Collections;
+import java.util.stream.Collectors;
+
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.jboss.tools.common.databinding.ObservablePojo;
+import org.jboss.tools.common.launcher.core.model.Booster;
+import org.jboss.tools.common.launcher.core.model.Catalog;
+import org.jboss.tools.common.launcher.core.model.Mission;
+
+public class NewLauncherProjectModel extends ObservablePojo {
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectModel.java]: a test for NewLauncherProjectModel would be great

File: common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardController.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	private IStatus getZip(IProgressMonitor monitor) {
+		CatalogManager manager = CatalogManager.getDefault();
+		ByteArrayOutputStream output = new ByteArrayOutputStream();
+		IStatus status;
+		try {
+			status =  manager.zip(LauncherCorePlugin.getDefault().getDefaultEndpointURL(), model.getSelectedBooster().getMission(), model.getSelectedBooster().getRuntime(), model.getSelectedBooster().getVersion(), model.getProjectName(), model.getGroupId(), model.getArtifactId(), model.getVersion(), output, monitor);
+			if (!monitor.isCanceled() && status.isOK()) {
+				status = unzip(output.toByteArray(), model.getLocation());
+			}
+		} finally {
```
Comment (adietish) [inline common/plugins/org.jboss.tools.common.launcher.ui/src/main/java/org/jboss/tools/common/launcher/ui/wizard/NewLauncherProjectWizardController.java]: I would think that there's no need for this finally block

**Comment by jeffmaury** [comment]
My remarks:

- I choose base because it seems to me the logical place to be and m2e dependency is not an issue because it's not the default in Eclipse distributions

- Updated page description and added tooltips for mission and runtime
- the error is raised by data binding and the RequiredControlDecorationUpdater
- the wizard is not Maven specific (think of NodeJS) and the M2E Maven wizard does not validate the version string
- Regarding the groovy JAR I also add issues with dependencies because it seems my M2 repo was garbage due to Nexus down on Aug 4th. Once I cleaned the failed dependencies it was ok for me

**Comment by adietish** [comment]
@jeffmaury 
> the error is raised by data binding and the RequiredControlDecorationUpdater

Looks like you're using the decorator that supports the "*" marker. This seems to indicate that you'd get the "*" decoration if you validate to CANCEL (instead of ERROR). Worth changing?

**Comment by jeffmaury** [comment]
The ERROR is generated by org.jboss.tools.common.ui.databinding.MandatoryStringValidator which is a utility class. Do you think we can switch to CANCEL ?

**Comment by adietish** [comment]
we agreed that I'll add the suggested improvements in the next sprint. Things work for me. +1

**Pull request merged**